packages/react-reconciler/src/ReactFiberHooks.new.js

```diff
@@ -721,6 +735,84 @@ function createFunctionComponentUpdateQueue(): FunctionComponentUpdateQueue {
   };
 }
 
+function use<T>(usable: Usable<T>): T {
+  if (
+    usable !== null &&
+    typeof usable === 'object' &&
+    typeof usable.then === 'function'
+  ) {
+    // This is a thenable.
+    const thenable: Thenable<T> = (usable: any);
+
+    // Track the position of the thenable within this fiber.
+    const index = thenableIndexCounter;
+    thenableIndexCounter += 1;
+
+    switch (thenable.status) {
+      case 'fulfilled': {
+        const fulfilledValue: T = (thenable.value: any);
+        return fulfilledValue;
+      }
+      case 'rejected': {
+        const rejectedError: mixed = thenable.value;
+        throw rejectedError;
+      }
+      default: {
+        const prevThenableAtIndex = getSuspendedThenable(index);
+        if (prevThenableAtIndex !== null) {
+          switch (prevThenableAtIndex.status) {
+            case 'fulfilled': {
+              const fulfilledValue: T = (prevThenableAtIndex.value: any);
+              return fulfilledValue;
+            }
+            case 'rejected': {
+              const rejectedError: mixed = prevThenableAtIndex.value;
+              throw rejectedError;
+            }
+            default: {
+              // The thenable still hasn't resolved. Suspend with the same
+              // thenable as last time to avoid redundant listeners.
+              throw prevThenableAtIndex;
+            }
+          }
+        } else {
+          // This is the first time something has been used at this index.
+          // Stash the thenable at the current index so we can reuse it during
+          // the next attempt.
+          trackSuspendedThenable(thenable, index);
+
+          // Assume that if the status is `pending`, there's already a listener
+          // that will set it to `fulfilled` or `rejected`.
+          // TODO: We could do this in `pingSuspendedRoot` instead.
+          if (thenable.status !== 'pending') {
+            thenable.status = 'pending';
+            thenable.then(
+              fulfilledValue => {
+                if (thenable.status === 'pending') {
+                  thenable.status = 'fulfilled';
+                  thenable.value = fulfilledValue;
+                }
+              },
+              (error: mixed) => {
+                if (thenable.status === 'pending') {
+                  thenable.status = 'rejected';
+                  thenable.value = error;
+                }
+              },
+            );
+          }
+          throw thenable;
```

Comment:
Should we add ping listeners directly here and instead throw a placeholder value?

Reply:
Yeah I can do that